File: types/product.ts
export enum TrendCategory {
  BREAKOUT = "breakout",
  SUSTAINED = "sustained",
  DISCOUNT_DRIVEN = "discount_driven",
}

export enum ViabilityGrade {
  A_PLUS = "A+",
  A = "A",
  A_MINUS = "A-",
  B_PLUS = "B+",
  B = "B",
  B_MINUS = "B-",
  C_PLUS = "C+",
  C = "C",
  C_MINUS = "C-",
  D = "D",
  E = "E",
}

export interface ProductSnapshot {
  timestamp: string;
  price: number;
  original_price?: number;
  sold_count: number;
  rating: number;
  review_count: number;
  in_stock: boolean;
}

export interface Product {
  id: string;
  title: string;
  image_url: string;
  category: string;
  shop_id: string;
  shop_name: string;
  country?: string; // Product origin/target market (US, UK, CA, etc.)

  // Current data
  current_price: number;
  original_price?: number;
  sold_count: number;
  rating: number;
  review_count: number;
  in_stock: boolean;

  // URLs
  product_url: string;
  shop_url: string;

  // Trend metrics
  trend_score: number;
  orders_3d_delta: number;
  orders_7d_delta: number;
  price_discount_rate: number;
  reviews_3d_delta: number;
  velocity_3d: number;
  acceleration: number;

  // Classification
  trend_category?: TrendCategory;
  viability_grade?: ViabilityGrade;
  viability_summary?: string;
  confidence_score: number;

  // PVS Components
  pv_demand_momentum?: number;
  pv_sustained_interest?: number;
  pv_price_stability?: number;
  pv_saturation?: number;
  pv_sentiment?: number;

  // Commission
  commission_rate?: number;
  has_affiliate_program: boolean;

  // Metadata
  first_seen: string;
  last_updated: string;
  snapshot_count: number;
}

export interface ProductDetail extends Product {
  snapshots: ProductSnapshot[];
  price_history: Array<{ date: string; price: number }>;
  sales_history: Array<{ date: string; sold_count: number }>;
}

export interface Shop {
  id: string;
  name: string;
  url: string;
  rating: number;
  total_products: number;
  followers?: number;
  badges: string[];
  response_rate?: number;
  ship_on_time_rate?: number;
  top_products: string[];
  first_seen: string;
  last_updated: string;
}

File: backend/app/models/product.py
"""
Product data models for TikTok Shop products.
"""
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime
from enum import Enum


class TrendCategory(str, Enum):
    """Product trend categories."""
    BREAKOUT = "breakout"  # New spike in orders
    SUSTAINED = "sustained"  # Steady growth
    DISCOUNT_DRIVEN = "discount_driven"  # Price drop driven


class ViabilityGrade(str, Enum):
    """Product viability rating (A-E scale)."""
    A_PLUS = "A+"  # Outstanding opportunity
    A = "A"  # Excellent opportunity
    A_MINUS = "A-"  # Very good opportunity
    B_PLUS = "B+"  # Good opportunity
    B = "B"  # Above average
    B_MINUS = "B-"  # Decent opportunity
    C_PLUS = "C+"  # Fair opportunity
    C = "C"  # Average
    C_MINUS = "C-"  # Below average
    D = "D"  # Poor opportunity
    E = "E"  # Very poor opportunity


class ProductSnapshot(BaseModel):
    """Single snapshot of product data at a point in time."""
    timestamp: datetime = Field(default_factory=datetime.utcnow)
    price: float
    original_price: Optional[float] = None
    sold_count: int  # "X sold" number
    rating: float
    review_count: int
    in_stock: bool = True


class Product(BaseModel):
    """TikTok Shop product with trend data."""
    id: str  # TikTok product ID
    title: str
    image_url: str
    category: str
    shop_id: str
    shop_name: str

    # Current data
    current_price: float
    original_price: Optional[float] = None
    sold_count: int
    rating: float
    review_count: int
    in_stock: bool = True

    # URLs
    product_url: str
    shop_url: str

    # Trend metrics
    trend_score: float = 0.0
    orders_3d_delta: int = 0
    orders_7d_delta: int = 0
    price_discount_rate: float = 0.0
    reviews_3d_delta: int = 0
    velocity_3d: float = 0.0  # orders per day
    acceleration: float = 0.0  # change in velocity

    # Classification
    trend_category: Optional[TrendCategory] = None
    viability_grade: Optional[ViabilityGrade] = None
    viability_summary: str = ""  # Plain-English summary of PVS
    confidence_score: float = 0.0  # 0-1 confidence in trend data

    # PVS Components (0-100 percentiles)
    pv_demand_momentum: float = 0.0  # 35% - TrendScore percentile
    pv_sustained_interest: float = 0.0  # 20% - 28d orders trend
    pv_price_stability: float = 0.0  # 15% - Margin potential
    pv_saturation: float = 0.0  # 15% - Low duplicate listings
    pv_sentiment: float = 0.0  # 10% - Rating & review growth
    pv_seasonality: float = 0.0  # 5% - (Optional later)

    # Commission (if available)
    commission_rate: Optional[float] = None
    has_affiliate_program: bool = False

    # Metadata
    first_seen: datetime = Field(default_factory=datetime.utcnow)
    last_updated: datetime = Field(default_factory=datetime.utcnow)
    snapshot_count: int = 0  # Number of data points collected

    class Config:
        json_schema_extra = {
            "example": {
                "id": "1234567890",
                "title": "Wireless Bluetooth Earbuds",
                "image_url": "https://example.com/image.jpg",
                "category": "Electronics",
                "shop_id": "shop123",
                "shop_name": "Tech Store",
                "current_price": 29.99,
                "original_price": 49.99,
                "sold_count": 15420,
                "rating": 4.7,
                "review_count": 3421,
                "product_url": "https://tiktok.com/@shop/product/1234567890",
                "shop_url": "https://tiktok.com/@shop",
                "trend_score": 87.5,
                "orders_3d_delta": 1250,
                "velocity_3d": 416.7,
                "trend_category": "breakout",
                "viability_grade": "A"
            }
        }


class ProductDetail(Product):
    """Extended product model with historical data."""
    snapshots: List[ProductSnapshot] = []
    price_history: List[dict] = []  # {date, price}
    sales_history: List[dict] = []  # {date, sold_count}


class Shop(BaseModel):
    """TikTok Shop seller information."""
    id: str
    name: str
    url: str
    rating: float
    total_products: int
    followers: Optional[int] = None
    badges: List[str] = []
    response_rate: Optional[float] = None
    ship_on_time_rate: Optional[float] = None

    # Top products
    top_products: List[str] = []  # Product IDs

    # Metadata
    first_seen: datetime = Field(default_factory=datetime.utcnow)
    last_updated: datetime = Field(default_factory=datetime.utcnow)
